### ISSUE TITLE
UnboundLocalError("local variable 'values' referenced before assignment") 

### ISSUE THREAD

**Issue by zabir-nabil** [PAIR-code/what-if-tool#216]
I have the training dataframe which has feature columns, preds (predicted score) and label column (ground truth). There is no nan values, all of them are float64.

I convert the dataframe to tf examples according to the reference notebook provided by what-if,

```
# Converts a dataframe into a list of tf.Example protos.
import tensorflow as tf
def df_to_examples(df, columns=None):
    examples = []
    if columns == None:
        columns = df.columns.values.tolist()
    for index, row in df.iterrows():
        example = tf.train.Example()
        for col in columns:
            if df[col].dtype is np.dtype(np.int64):
                example.features.feature[col].int64_list.value.append(int(row[col]))
            elif df[col].dtype is np.dtype(np.float64):
                example.features.feature[col].float_list.value.append(row[col])
            else:
              example.features.feature[col].float_list.value.append(row[col]) # just float64 instead of numpy.float64
        examples.append(example)
    return examples
```
```
num_datapoints = 2000
test_examples = df_to_examples(test_features[0:num_datapoints])
```

```
config_builder = WitConfigBuilder(
    test_examples,
    feature_names=features + ['label', 'preds']).set_custom_predict_fn(
  model.predict).set_target_feature('label')
```

Once the tool is launched, and I try to click on the predict button, I get the following error:
UnboundLocalError("local variable 'values' referenced before assignment") 

Screenshot: https://ibb.co/fG3STcP

**Comment by jameswex** [comment]
The issue is most likely in your custom prediction fn. Perhaps you can write your predict function with logging to see exactly what is passed in, if its in the format your model expects, or if you need to preprocess the data sent to the function before you pass it to your model.

**Comment by zabir-nabil** [comment]
Solved it. Thanks for the input. Now, I have written another custom_function which takes a train.Example and performs the predict, however I'm getting a new error `TypeError("object of type 'float' has no len()")`

My custom predict function:

```
def custom_predict_fn(examples):
  if type(examples) == list:
    feats_all = []
    for example in examples:
      feats = []
      for col in features:
        #print(col)
        p = example.features.feature[col].float_list.value
        #print(p)
        feats.append(p[0])
      feats_all.append(feats)
  else:
    feats = []
    for col in features:
      #print(col)
      p = examples.features.feature[col].float_list.value
      #print(p)
      feats.append(p[0])
    feats_all = [feats]
  return model.predict(np.array(feats_all)).tolist()
```

This is how my train.Example sample looks like:

```
features {
  feature {
    key: "a1"
    value {
      float_list {
        value: 1.0
      }
    }
  }
  feature {
    key: "a2"
    value {
      float_list {
        value: 1.0
      }
    }
  }
  feature {
    key: "a3"
    value {
      float_list {
        value: 1.0
      }
    }
  }
  feature {
    key: "cj_val"
    value {
      float_list {
        value: -0.9179026484489441
      }
    }
  }
  feature {
    key: "cos"
    value {
      float_list {
        value: -0.561825692653656
      }
    }
  }
  feature {
    key: "hr"
    value {
      float_list {
        value: 35.0
      }
    }
  }
  feature {
    key: "iscore"
    value {
      float_list {
        value: 8.0
      }
    }
  }
  feature {
    key: "label"
    value {
      float_list {
        value: 0.0
      }
    }
  }
  feature {
    key: "max_pct_2"
    value {
      float_list {
        value: 48.029998779296875
      }
    }
  }
  feature {
    key: "max_pct"
    value {
      float_list {
        value: 0.4039580523967743
      }
    }
  }
```

**Comment by jameswex** [comment]
Nice to see that the function works with your manual testing. Are you able to log the input to the custom predict fn when call from the UI to see how it differs from your manual testing?

**Comment by zabir-nabil** [comment]
Yeah, I just saw that, it looks same as the `custom_predict_fn(test_examples[2:5])` version, an array of examples. It logged everything fine, I tried to catch any Exception too, but nothing was shown, so not sure why it gives me the error.

**Comment by jameswex** [comment]
it may be the case that if you pass a single example through the ui (by pressing predict button after making an edit), then your custom predict fn is returning a raw values instead of a list of length 1 of the value to return?

**Comment by zabir-nabil** [comment]
Okay, interesting. I have tested it with single example, it still returns a list, something like [0.4]. I even added a check just before the return value, it seems no error is taking place there; the error is taking place after the value is returned most probably. Is there any specific format that the UI expects internally? I seems it's trying to get a length of a float value, maybe a feature?

**Comment by zabir-nabil** [comment]
Okay, it's working now. From the comments, I found out the prediction should have shape like (batch, num_classes) for classification, but my model was predicting a single score for a binary classification problem. Now, it's working as expected. Thanks @jameswex

**Comment by jameswex** [comment]
Great to hear!
